### PR TITLE
add:cmake: add compilation flags to harden the security of navit

### PR DESCRIPTION
**Pull request by aerostitch**
Hi guys,

This PR adds a few compilation flags that detects issues in advance.
Those are generally set in Debian as best practices and right now, those return errors (see #966 ).
So I propose we add them to make sure that we can detect those issues in advance.

What do you think?
Thanks
Joseph

### CODE REVIEW

**Comment by aerostitch** [comment]
Not sure yet why it's not failing! :\

File: CMakeLists.txt [hunk 1]
```diff
@@ -836,6 +836,10 @@ if(CMAKE_COMPILER_IS_GNUCC OR CCMAKE_COMPILER_IS_GNUCXX)
 	add_compiler_flag_if_available("-Wmissing-prototypes")
 	add_compiler_flag_if_available("-Wstrict-prototypes")
 	add_compiler_flag_if_available("-Wformat-security")
+	add_compiler_flag_if_available("-Werror=format-security")
```
Comment (metalstrolch) [inline CMakeLists.txt]: At least on my Gentoo setup this does nothing as something set to CMAKE_C_FLAGS or CMAKE_CXX_FLAGS at line 831 does cause the check_c_compiler_flag macro used inside the add_compiler_flag_if_available macro to always fail. So the other compiler flags are not in use at all currently.

**Comment by aerostitch** [comment]
So I removed the `if(CMAKE_COMPILER_IS_GNUCC OR CCMAKE_COMPILER_IS_GNUCXX)` altogether as it's not the recommended way to test the compiler according to https://cmake.org/cmake/help/latest/variable/CMAKE_COMPILER_IS_GNUCC.html and it's also only available starting in cmake 3.7 and we require only cmake 3.2.

Plus changing the add of default flags to all use our `add_compiler_flag_if_available` macro so that there's only 1 way of adding flags which seems simpler.
Do those flags show up now on your Gentoo?

It should look something like that:
```
-- Performing Test FLAG_AVAILABLE_C_-Wdate-time
-- Performing Test FLAG_AVAILABLE_C_-Wdate-time - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wdate-time
-- Performing Test FLAG_AVAILABLE_CXX_-Wdate-time - Success
```

File: CMakeLists.txt (hunk 1 above)
Comment (metalstrolch) [reply CMakeLists.txt]: Gentoo is using gcc 9.2.0 today. And on my box it entered the if. CMAKE even tells you in the log how it evaluated the flags. This time it looks better. Te problem is that existing flags in the current variables affect the check if the flag exists causing false negatives.

No clue how this does using different compiler. But I think even windows has gcc has it?

```
-- Performing Test FLAG_AVAILABLE_C_-Wall
-- Performing Test FLAG_AVAILABLE_C_-Wall - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wall
-- Performing Test FLAG_AVAILABLE_CXX_-Wall - Success
-- Performing Test FLAG_AVAILABLE_C_-Wno-unused-parameter
-- Performing Test FLAG_AVAILABLE_C_-Wno-unused-parameter - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wno-unused-parameter
-- Performing Test FLAG_AVAILABLE_CXX_-Wno-unused-parameter - Success
-- Performing Test FLAG_AVAILABLE_C_-Wno-sign-compare
-- Performing Test FLAG_AVAILABLE_C_-Wno-sign-compare - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wno-sign-compare
-- Performing Test FLAG_AVAILABLE_CXX_-Wno-sign-compare - Success
-- Performing Test FLAG_AVAILABLE_C_-Wno-missing-field-initializers
-- Performing Test FLAG_AVAILABLE_C_-Wno-missing-field-initializers - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wno-missing-field-initializers
-- Performing Test FLAG_AVAILABLE_CXX_-Wno-missing-field-initializers - Success
-- Performing Test FLAG_AVAILABLE_C_-Wundef
-- Performing Test FLAG_AVAILABLE_C_-Wundef - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wundef
-- Performing Test FLAG_AVAILABLE_CXX_-Wundef - Success
-- Performing Test FLAG_AVAILABLE_C_-Wcast-align
-- Performing Test FLAG_AVAILABLE_C_-Wcast-align - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wcast-align
-- Performing Test FLAG_AVAILABLE_CXX_-Wcast-align - Success
-- Performing Test FLAG_AVAILABLE_C_-Wpointer-arith
-- Performing Test FLAG_AVAILABLE_C_-Wpointer-arith - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wpointer-arith
-- Performing Test FLAG_AVAILABLE_CXX_-Wpointer-arith - Success
-- Performing Test FLAG_AVAILABLE_C_-Wextra
-- Performing Test FLAG_AVAILABLE_C_-Wextra - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wextra
-- Performing Test FLAG_AVAILABLE_CXX_-Wextra - Success
-- Performing Test FLAG_AVAILABLE_C_-Wdate-time
-- Performing Test FLAG_AVAILABLE_C_-Wdate-time - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wdate-time
-- Performing Test FLAG_AVAILABLE_CXX_-Wdate-time - Success
-- Performing Test FLAG_AVAILABLE_C_-Wmissing-prototypes
-- Performing Test FLAG_AVAILABLE_C_-Wmissing-prototypes - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wmissing-prototypes
-- Performing Test FLAG_AVAILABLE_CXX_-Wmissing-prototypes - Failed
-- Performing Test FLAG_AVAILABLE_C_-Wstrict-prototypes
-- Performing Test FLAG_AVAILABLE_C_-Wstrict-prototypes - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wstrict-prototypes
-- Performing Test FLAG_AVAILABLE_CXX_-Wstrict-prototypes - Failed
-- Performing Test FLAG_AVAILABLE_C_-Wformat-security
-- Performing Test FLAG_AVAILABLE_C_-Wformat-security - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Wformat-security
-- Performing Test FLAG_AVAILABLE_CXX_-Wformat-security - Success
-- Performing Test FLAG_AVAILABLE_C_-Werror=format-security
-- Performing Test FLAG_AVAILABLE_C_-Werror=format-security - Success
-- Performing Test FLAG_AVAILABLE_CXX_-Werror=format-security
 
```

Comment (aerostitch) [reply CMakeLists.txt]: Yeah via mingwin or something like that (I haven't touched a windows in many years so not 100% sure tbh).
So would that latest change work for you?

**Pull request merged**